### DISCUSSION TITLE
I2S Master clock output from PWM reliability?

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#11866] [Libraries & Drivers] [answered]
Hi,

I'm experimenting with an ADC (PCM1808) and a codec (PCM3060) which both require a master system clock when in slave mode (currently only slave mode is supported as I understand from the MicroPython docs about [I2S](https://docs.micropython.org/en/latest/library/machine.I2S.html#machine-i2s)). As mentioned in other forum/discussion posts, it should possible to generate that clock by running a PWM signal on a GPIO.
I tested this on an ESP32S3, Nucleo-STM32H723 and a pyboard v1.1 but it seems that some jitter or sync issue might occur. I the MCK from a PWM reliable enough and is it in "sync" with the I2S generated clocks (SCK/BCK, WS). Is there a way to enable this MCK like in the iMXRT port? Or has someone ran a codec or adc/dac as slave mode with a PWM as I2S master clock successfully?
Thanks

**Comment by robert-hh** [comment, marked as answer]
The PWM can be used for MCK generation. When I used it, the MCK clocked the internal logic of the codec and had not to be in sync with the audio clocks. The ESP32 PWM can create stable frequencies, if the ratio between the ESP32 internal 160Mhz and the PWM signal is an integer. 8MHz is fine, 12Mhz will be a mix of two frequencies.

8MHz PWM:
![pwm_8MHz](https://github.com/micropython/micropython/assets/12476868/4bea1386-bdb1-4520-a76a-ef28f6ead0bd)

12 MHZ PWM
![pwm_12MHz](https://github.com/micropython/micropython/assets/12476868/23ed93c2-da95-46f2-89b3-00807209fa3e)

**Reply by saraverbeecke** [reply]
Thanks for testing this Robert. I had some noise on the GND that made the clock jitter too much. I can now use a PWM as clock for the ADC.
